File: leetcode/src/main/problems/array/AnagramValid.java
package array;

import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;

/*
    https://leetcode.com/problems/valid-anagram/

    Anagrams là 2 chuỗi s1, s2 có cùng những ký tự (bằng nhau sau khi sắp xếp)

    Xem thêm GroupAnagrams.java
 */
public class AnagramValid {

    public boolean isAnagram_1(String s1, String s2) {
        if (s1.length() != s2.length()) return false;
        char[] mark = new char[26];
        for (int i = 0; i < s1.length(); i++) {
            mark[s1.charAt(i) - 'a']++;
            mark[s2.charAt(i) - 'a']--;
        }
        for (char v : mark) {
            if (v != 0) return false;
        }
        return true;
    }

    public boolean isAnagram_2(String s1, String s2) {
        if (s1.length() != s2.length()) return false;
        int[] mark = new int[26];
        s1.chars().forEach(c -> mark[c - 'a']++);
        s2.chars().forEach(c -> mark[c - 'a']--);
        return Arrays.stream(mark).allMatch(v -> v == 0);
    }

    public boolean isAnagram(String s1, String s2) {
        if (s1.length() != s2.length()) return false;
        Map<Character, Integer> counter = new HashMap<>();

        // Increment character counts for s1
        for (char c : s1.toCharArray()) {
            counter.put(c, counter.getOrDefault(c, 0) + 1);
        }

        // Decrement character counts for s2
        for (char c : s2.toCharArray()) {
            counter.put(c, counter.getOrDefault(c, 0) - 1);
        }

        // Check if all character counts are zero
        return counter.values().stream().allMatch(v -> v == 0);
    }

    @Test
    public void test() {
        assertEquals(true, isAnagram("anagram", "nagaram"));
        assertEquals(false, isAnagram("rat", "car"));
    }

}

File: go_algorithm/arrays/anagram_valid.go
package main

import "go_algorithm/test_util"

/*
https://leetcode.com/problems/valid-anagram/

Anagrams là hai chuỗi s1, s2 có cùng những ký tự (bằng nhau sau khi sắp xếp).
*/
func isAnagram_OK(s1 string, s2 string) bool {
	mark := [26]int{}
	// Mark s1
	for _, c := range s1 {
		mark[c-'a']++
	}
	// Mark s2
	for _, c := range s2 {
		mark[c-'a']--
		// Return sớm
		if mark[c-'a'] < 0 {
			return false
		}
	}
	// Check
	for _, v := range mark {
		if v != 0 {
			return false
		}
	}

	return true
}

func isAnagram(s1, s2 string) bool {
	if len(s1) != len(s2) {
		return false
	}

	var count [26]int
	for i := range s1 {
		count[s1[i]-'a']++
		count[s2[i]-'a']--
	}

	for _, v := range count {
		if v != 0 {
			return false
		}
	}

	return true
}

func main() {
	test_util.AssertBool(true, isAnagram("anagram", "nagaram"))
	test_util.AssertBool(false, isAnagram("rat", "car"))
}
